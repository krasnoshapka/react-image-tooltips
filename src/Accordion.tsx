import * as React from 'react';
import { Spring } from 'react-spring/renderprops.cjs';
import useFirstMount from './hooks/firstmount';

export interface AccordionProps extends React.ComponentPropsWithoutRef<"div"> {
  className?: string;
  children?: React.ReactNode;
  toggled?: boolean;
  onFullyShown?: () => void;
  onFullyHidden?: () => void;
}

export const Accordion: React.FC<AccordionProps> =({
  className = '',
  children,
  toggled = false,
  onFullyShown,
  onFullyHidden,
  ...props
}: AccordionProps) => {
  const isFirstMount = useFirstMount();

  const handleAnimationRest = React.useCallback(() => {
    if (toggled && onFullyShown) onFullyShown();
    else if (!toggled && onFullyHidden) onFullyHidden();
  }, [toggled, onFullyShown, onFullyHidden]);

  const cssClass = ["hotspot-content", className];

  return (
    <Spring
      config={{
        tension: 170,
        friction: 22,
        clamp: true
      }}
      from={{ height: 0 }}
      to={{ height: toggled ? 'auto' : 0 }}
      immediate={isFirstMount}
      onRest={handleAnimationRest}
    >
      {styleProps => (
        <div className={cssClass.join(' ')} style={styleProps} {...props}>
          {children}
        </div>
      )}
    </Spring>
  );
}
