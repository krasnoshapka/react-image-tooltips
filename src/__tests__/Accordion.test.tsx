import * as React from 'react';
import {render} from '@testing-library/react';
import {Accordion} from '../Accordion';

describe('<Accordion />', () => {
  it('renders correctly when open', () => {
    const {container} = render(
      <Accordion toggled={true}>
        <span>Accordion content</span>
      </Accordion>
    );

    expect(container).toMatchSnapshot();
  });

  it('renders correctly when closed', () => {
    const {container} = render(
      <Accordion toggled={false}>
        <span>Accordion content</span>
      </Accordion>
    );

    expect(container).toMatchSnapshot();
  });

  it('appends passed classnames', () => {
    const {container} = render(
      <Accordion className="test-accordion">Children</Accordion>
    );

    expect(container.firstChild).toHaveClass('test-accordion');
  });

  it('forwards extra props', () => {
    const {container} = render(
      <Accordion id="test-id">Children</Accordion>
    );

    expect(container.firstChild).toHaveAttribute('id', 'test-id');
  });

  it('should open', () => {
    const {container, rerender} = render(
      <Accordion toggled={false}>Children</Accordion>
    );

    rerender(<Accordion toggled={true}>Children</Accordion>);

    expect(container.firstChild).toHaveStyle({ height: 'auto' });
  });

  it('calls onFullyShown', () => {
    const onFullyShown = jest.fn();

    const {rerender} = render(
      <Accordion toggled={false} onFullyShown={onFullyShown}>
        Children
      </Accordion>
    );

    expect(onFullyShown).not.toBeCalled();

    rerender(
      <Accordion toggled={true} onFullyShown={onFullyShown}>
        Children
      </Accordion>
    );

    expect(onFullyShown).toBeCalled();
  });

  it('calls onFullyHidden', () => {
    const onFullyHidden = jest.fn();

    const {rerender} = render(
      <Accordion toggled={true} onFullyHidden={onFullyHidden}>
        Children
      </Accordion>
    );

    expect(onFullyHidden).not.toBeCalled();

    rerender(
      <Accordion toggled={false} onFullyHidden={onFullyHidden}>
        Children
      </Accordion>
    );

    expect(onFullyHidden).toBeCalled();
  });
});
