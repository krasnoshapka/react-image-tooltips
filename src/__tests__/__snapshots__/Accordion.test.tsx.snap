// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Accordion /> renders correctly when closed 1`] = `
<div>
  <div
    class="hotspot-content "
    style="height: 0px;"
  >
    <span>
      Accordion content
    </span>
  </div>
</div>
`;

exports[`<Accordion /> renders correctly when open 1`] = `
<div>
  <div
    class="hotspot-content "
    style="height: auto;"
  >
    <span>
      Accordion content
    </span>
  </div>
</div>
`;
